pennylane/resource/resource.py

```diff
@@ -0,0 +1,169 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+The data class which will aggregate all the resource information from a quantum workflow.
+"""
+
+
+class Resources:
+    r"""Contains attributes which store key resources such as number of gates, number of wires, shots,
+    depth and gate types, tracked over a quantum workflow.
+
+    Args:
+        num_wires (int): number of qubits
+        num_gates (int): number of gates
+        gate_types (dict): dictionary storing operation names (str) as keys
+            and the number of times they are used in the circuit (int) as values
+        depth (int): the depth of the circuit defined as the maximum number of non-parallel operations
+        shots (int): number of samples to generate
+
+    Raises:
+        TypeError: If the attributes provided are not of the correct type.
+        ValueError: If an attribute provided has value less than 0.
+        AttributeError: If an attribute is set after initialization.
+
+    .. details::
+
+        The resources being tracked can be accessed and set as class attributes.
+        Additionally, the :code:`Resources` instance can be nicely displayed in the console.
+
+        **Example**
+
+        >>> r = Resources(num_wires=2, num_gates=2, gate_types={'Hadamard': 1, 'CNOT':1}, depth=2)
+        >>> print(r)
+        wires: 2
+        gates: 2
+        depth: 2
+        shots: 0
+        gate_types: {'Hadamard': 1, 'CNOT': 1}
+    """
+
+    def __init__(
+        self, num_wires=0, num_gates=0, gate_types={}, depth=0, shots=0
+    ):  # pylint: disable=too-many-arguments, dangerous-default-value
+        """Initialize a Resources instance and perform input type validation."""
+
+        if not all(isinstance(param, int) for param in [num_wires, num_gates, depth, shots]):
```

Comment:
I think it could be useful to check that summing the number of gates in `gate_types` adds up to `num_gates` and overwriting `self._num_gates` accordingly. Might not be necessary if this class isn't meant to be user facing.

Reply:
Great point! Thanks for clarifying.